### PR TITLE
Add a basic blocking QSPI interface

### PR DESCRIPTION
**Pull request by X-yl**
This module implements the QuadSPI interface which allows high speed communication with external flash memory.

Limitations:
    - Interrupts are not supported.
    - Status polling mode is not supported.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -154,6 +154,8 @@ pub mod prelude;
 #[cfg(feature = "device-selected")]
 pub mod qei;
 #[cfg(feature = "device-selected")]
+pub mod qspi;
```
Comment (burrbull) [inline src/lib.rs:157]:
```suggestion
#[cfg(feature = "quadspi")]
pub mod qspi;
```

File: src/qspi.rs [hunk 2]
```diff
@@ -0,0 +1,783 @@
...
+                .bit(command.double_data_rate)
+                .dcyc()
+                .bits(command.dummy_cycles)
+                .dmode()
+                .bits(command.data_mode as u8)
+                .imode()
+                .bits(imode)
+                .instruction()
+                .bits(instruction)
+        });
```
Comment (burrbull) [inline src/qspi.rs:586]: You can do this more compact
```suggestion
        self.qspi.ccr.modify(|_, w| unsafe {
            w.fmode().bits(0b11 /* Memory mapped mode */);
            w.admode().bits(command.address_mode as u8);
            w.adsize().bits(self.config.address_size as u8);
            w.abmode().bits(abmode);
            w.absize().bits(absize);
            w.ddrm().bit(command.double_data_rate);
            w.dcyc().bits(command.dummy_cycles);
            w.dmode().bits(command.data_mode as u8);
            w.imode().bits(imode);
            w.instruction().bits(instruction)
        });```

File: src/qspi.rs [hunk 3]
```diff
@@ -0,0 +1,783 @@
+//! QSPI interface
+//!
+//! This module implements the QuadSPI interface which allows high speed
+//! communication with external flash memory.
+//!
+//! Limitations:
+//!     - Interrupts are not supported.
```
Comment (burrbull) [inline src/qspi.rs:7]:
```suggestion
//!     - Quad mode only.
//!     - Interrupts are not supported.
```

File: src/qspi.rs [hunk 4]
```diff
@@ -0,0 +1,783 @@
...
+        alt::quadspi::Bk2Io3,
+        alt::quadspi::Clk,
+    );
+}
+
+pub struct Qspi<Pins: QspiPins> {
+    qspi: QUADSPI,
+    config: QspiConfig,
+    _marker: PhantomData<Pins>,
+}
```
Comment (burrbull) [inline src/qspi.rs:53]: I'd rename `Pins` -> `BANK` and keep pins inside. This allows you to add `release` for `Qspi`.
```suggestion
pub struct Qspi<BANK: QspiPins> {
    qspi: QUADSPI,
    config: QspiConfig,
    pins: BANK::Pins,
}
```

**Comment by X-yl** [comment]
Thanks for your suggestions, I've amended the commit to incorporate them. Let me know if any other changes are required :)

**Comment by X-yl** [comment]
Added `#[cfg(feature = "quadspi")]` to the RccBus stuff.

**Comment by burrbull** [comment]
Could you add example?

**Comment by burrbull** [comment]
> Thanks for your suggestions, I've amended the commit to incorporate them. Let me know if any other changes are required :)

https://github.com/stm32-rs/stm32f4xx-hal/commit/6650964c084cb8f3cb5d6ad8e006f22bcfb4b44b

**Comment by X-yl** [comment]
Applied your change and added an example using a W25Q series flash chip.

**Comment by X-yl** [comment]
(Adjusted clock prescaler for max speed)

**Comment by burrbull** [comment]
Is `Speed::VeryHigh` required by QSPI / default?

**Comment by X-yl** [comment]
> Is `Speed::VeryHigh` required by QSPI / default?

I don't think it's required, but that was the configuration generated by CubeMX so I used it as is. From my testing, VeryHigh is required for prescaler zero but it works without setting the speed for higher prescaler.

**Comment by burrbull** [comment]
> I don't think it's required, but that was the configuration generated by CubeMX so I used it as is. From my testing, VeryHigh is required for prescaler zero but it works without setting the speed for higher prescaler.

https://github.com/stm32-rs/stm32f4xx-hal/commit/516a0091a21f96bdb70825398754d364aeb42563

**Comment by X-yl** [comment]
Fixed clippy suggestions

File: examples/qspi-w25q.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+pub struct W25Q<PINS: QspiPins> {
+    qspi: Qspi<PINS>,
+}
+
+pub struct DeviceId(u8);
+
+impl<PINS> W25Q<PINS>
+where
+    PINS: QspiPins + 'static,
```
Comment (burrbull) [inline examples/qspi-w25q.rs:25]:
```suggestion
    PINS: QspiPins,
```

**Review by burrbull — APPROVED** [review]

**Pull request merged**